src/extension/extensionServer.ts

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for details.
+
+import * as em from "../common/extensionMessaging";
+import * as http from "http";
+import * as vscode from "vscode";
+import * as Q from "q";
+
+export class ExtensionServer implements vscode.Disposable {
+
+    private serverInstance: http.Server = null;
+
+    /**
+     * Starts the server.
+     */
+    public setup(): Q.Promise<void> {
+
+        let deferred = Q.defer<void>();
+
+        let requestHandler = (request: http.IncomingMessage, response: http.ServerResponse) => {
+            this.handleIncomingMessage(request, response);
+        };
+
+        let launchCallback = (error: any) => {
+            if (error) {
+                deferred.reject(error);
+            } else {
+                deferred.resolve(null);
+            }
+        };
+
+        this.serverInstance = http.createServer(requestHandler);
+        this.serverInstance.listen(em.ServerParams.PORT, null, null, launchCallback);
+
+        return deferred.promise;
+    }
+
+    /**
+     * Stops the server.
+     */
+    public dispose(): void {
+        if (this.serverInstance) {
+            this.serverInstance.close();
+        }
```

Comment:
I'd assign null to serverInstance after we close it, just to have 2 possible states in the object (null or non-null and opened).